tests/phpunit/tests/test-flags.php

```diff
@@ -20,17 +22,24 @@ static function wpTearDownAfterClass() {
 	}
 
 	function setUp() {
-		self::$polylang = new PLL_Frontend( self::$polylang->links_model );
+		parent::setUp();
+
+		$options       = array_merge( PLL_Install::get_default_options(), array( 'default_lang' => 'en_US' ) );
+		$model         = new PLL_Model( $options );
+		$links_model   = new PLL_Links_Default( $model );
+		$this->pll_env = new PLL_Frontend( $links_model );
+		$this->pll_env->init();
+		$this->pll_env->model->cache->clean();
```

Comment:
Is it useful? You just instantiated the PLL_Model object. So I expect that the cache is clean. Am I Wrong?

Reply:
In fact, if you remove this line, one test fails. Because any previous call to `PLL_Model::get_language()` would have set values in the cache, we would otherwise be checking values set from the `PLL_Frontend::init()` values above rather than those computed in the test themselves...

On another hand, it doesn't seem that instantiating and initializing a `PLL_Frontend` object is mandatory for these tests to pass. Although, I am not sure were the code that is supposed to make these tests pass resides, so I am not able to check whether setting a `PLL_Frontend` instance is relevant or not.

Reply:
I thought that I already answered to this. I may have forgotten to push the green button :(

It looks to me that since the tests use only the `PLL_Model` instance, we have no need for the `PLL_Frontend` and `PLL_Links_Default` instances.

Reply:
In fact, a lot of the logic tested here happens in [PLL_Links_Model::pll_languages_list()](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/include/links-model.php#L98-L104) and [PLL_Links_Model::pll_after_languages_cache()](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/include/links-model.php#L115-L126), which are triggered on [pll_languages_list](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/include/model.php#L89) and [pll_after_languages_cache](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/include/model.php#L116).

These filters are registered during the [instantiation of the PLL_Links_Model subclass](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/include/links-model.php#L29-L30) in [PLL_Model::get_links_model()](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/include/model.php#L594), that happens twice:

- During the tests bootstrap, in [Polylang::init()](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/include/class-polylang.php#L192), which is triggered by WordPRess `plugins_loaded` hook.
- In [PLL_UnitTestCase::wpSetUpBeforeClass()](https://github.com/polylang/polylang/blob/5d2a023bf06e0159022a1e716a4df5d027d880fc/tests/phpunit/includes/testcase-trait.php#L23)

Both filters are backed up in [WP_UnitTestCase_Base](https://github.com/WordPress/wordpress-develop/blob/3ebeaffe80d31511a0de7f0aa37fb181d9832e82/tests/phpunit/includes/abstract-testcase.php#L111), then triggered in each test.

The commit [c22d9990b](https://github.com/polylang/polylang/pull/566/commits/c22d90b41d428df0a1d1cacc4953b4bc426149ab) removes both of these hooks to use some set in the test `setUp()` method instead.